src/js/libcs/OpDrawing.js

```diff
@@ -520,168 +534,131 @@ eval_helper.drawconic = function(conicMatrix, modifs) {
+            }
+            coord = sort > 0 ? sol[1] : sol[0];
+            if (0 < coord && coord < extent || coord === extent0) {
+                bInside = !bInside; // toggles at an intersection
+                t = bInside ? "begin" : "end";
+                pt = vert ? mkp(other, coord, t) : mkp(coord, other, t);
+                points.push(pt);
+            }
+        } else {
+            var x, y;
+            var flat = flats[vert ? 0 : 1];
+            if (flat !== null) {
+                t = "split";
+                flat = flat[other === 0 ? 0 : 1];
+                if (vert) { // xFlat for vertical tangent to oval
+                    x = flat;
+                    if (0 <= x && x <= csw) {
+                        y = -0.5 * (c11 * x + c01) / c02;
+                        if (0 <= y && y <= csh) points.push(mkp(x, y, t));
+                    }
+                } else { // !vert => yFlat for horizontal tangent to oval
+                    y = flat;
+                    if (0 <= y && y <= csh) {
+                        x = -0.5 * (c11 * y + c10) / c20;
+                        if (0 <= x && x <= csw) points.push(mkp(x, y, t));
+                    }
+                }
+            }
         }
         return true;
     }
 
-    if (!(doBoundary(tl, bl, yLeft, 0, 0, 1, 1, csh) &&
-            doBoundary(bl, br, xBottom, 0, csh, 0, 1, csw) &&
-            doBoundary(tr, br, yRight, csw, 0, 1, -1, csh) &&
-            doBoundary(tl, tr, xTop, 0, 0, 0, -1, csw)))
+    if (!(doBoundary(yLeft, 0, true, +1, csh) &&
+            doBoundary(xBottom, csh, false, +1, csw) &&
+            doBoundary(yRight, csw, true, -1, csh) &&
+            doBoundary(xTop, 0, false, -1, csw)))
         return;
 
-    function segmentEllipse() {
-        if (discr <= 0)
-            return false;
-
-        var initialLength = boundary.length;
-
-        // Compute the roots of the y discriminant
-        // for points with vertical tangents
-        sol = solveRealQuadratic(k00, -2 * k10, k20);
-        if (sol) {
-            for (i = 0; i < 2; ++i)
-                if (sol[i] >= 0 && sol[i] <= csw)
-                    boundary.push(
-                        mkp(sol[i], -0.5 * (c11 * sol[i] + c01) / c02));
+    csctx.beginPath();
+    var next;
+    var n = points.length;
+    if (n < 2) return; // Nothing to draw
+    var i = n - 1;
+    var previous = points[i]; // Wrap-around
+    var previousBegin = null;
+    do {
+        next = points[i];
+        if (next.t === "begin") {
+            previousBegin = next;
+            break;
         }
-
-        // Compute the roots of the x discriminant
-        // for points with horizontal tangents
-        sol = solveRealQuadratic(k00, -2 * k01, k02);
-        if (sol) {
-            for (i = 0; i < 2; ++i)
-                if (sol[i] >= 0 && sol[i] <= csh)
-                    boundary.push(
-                        mkp(-0.5 * (c11 * sol[i] + c10) / c20, sol[i]));
+    } while (--i > 0);
```

Comment:
I consider the `do … while` harder to read than a corresponding `for`, so I'd suggest

```js
for (i = 0; i < n; ++i)
    if (points[i].t === "begin")
        previousBegin = points[i];
```

If forward iteration over these few points feels too wasteful, or you want to keep two levels of braces and the `next` variable for the sake of clarity, at least make it `for (i = n - 1; i >= 0; --i)`.

I do have to admit that I was worried for a moment about the fact that you excluded the `i === 0` case. Makes sense reading subsequent code, but feels like a bug at first. So although it would be fine to write the terminal condition as `i > 0` resp. the initialization as `i = 1` in the code above, I'd rather not do that to avoid the confusion.